File: 강의 자료/02-알고리즘/12-트리/문제별 코드/11725-트리의 부모 찾기/solution.py
import sys
from collections import deque
si = sys.stdin.readline
n = int(si())
adj = [[] for _ in range(n + 1)]
parent = [0] * (n + 1)
sys.setrecursionlimit(100005)

for _ in range(n - 1):
    x, y = list(map(int, si().split()))
    adj[x].append(y)
    adj[y].append(x)

def dfs(x, par):
    for y in adj[x]:
        if y == par:
            continue
        parent[y] = x
        dfs(y, x)

dfs(1, -1)
for i in range(2, n + 1):
    print(parent[i])

File: 강의 자료/02-알고리즘/12-트리/문제별 코드/11725-트리의 부모 찾기/solution.cpp
#include <iostream>
#include <algorithm>
#pragma warning(disable:4996)
using namespace std;
#define NM 100005

#include <vector>
vector<int> adj[NM];
int n, m, parent[NM];
void input() {
	cin >> n;
	for (int i = 1; i < n; i++) {
		int u, v;
		cin >> u >> v;
		adj[u].push_back(v);
		adj[v].push_back(u);
	}
}

bool visit[NM];
void DFS(int x, int par) {
	for (int y: adj[x]){
		if (y == par) continue;
		parent[y] = x;
		DFS(y, x);
	}
}

void pro(){
	DFS(1, -1);
	for (int i = 2; i <= n; i++) {
		cout << parent[i] << '\n';
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	input();
	pro();
	return 0;
}